File: 03 Classification/03 imbalanced outcome.R
library(tidymodels)
# library(kknn)

# The Data & Problem --------------------------------------------------------

data("Caravan", package = "ISLR")
?ISLR::Caravan

levels(Caravan$Purchase)
# We will relevel the factor so the *first* class is the event class!
Caravan$Purchase <- relevel(Caravan$Purchase, ref = "Yes")
levels(Caravan$Purchase)
# This way we won't have to set `event_level = "second"` everywhere.

# Data Splitting
set.seed(1234)
splits <- initial_split(Caravan, prop = 0.7, strata = Purchase)
Caravan.train <- training(splits)
Caravan.test <- testing(splits)


table(Caravan.train$Purchase) |> proportions()
# As we can see, the classes are very unbalanced.

# Note we used a stratified split, so that both the train and test set have
# about the ~same distribution of classes. This is particularly important with
# imbalanced data.
table(Caravan.test$Purchase) |> proportions()


# We will be using this recipe:
rec <- recipe(Purchase ~ ., data = Caravan.train) |>
  step_normalize(all_numeric_predictors())

# And these classification metrics:
mset_class <- metric_set(accuracy, specificity, sensitivity)


# The worst model ---------------------------------------------------------
# The data imbalance means that, technically, we can achieve high accuracy by
# simply predicting "No"....

null_fit <-
  workflow(preprocessor = rec, spec = null_model(mode = "classification")) |>
  fit(data = Caravan.train)
# The null model is the worst possible model - it does not use any information
# in X, only the distribution of Y in the training set.
# - For regression, it always predicts mean(Y)
# - For classification, it predicts the frequent class and base rate
#   probabilities.
# These models are good for benchmarking.

# But like a "real" model, it can be used to generate predictions on new data:
Caravan.test_predictions_NULL <- augment(null_fit, new_data = Caravan.test)

# Same as:
# Caravan.test$.pred_class.BAD <- factor("No", levels = c("Yes", "No"))
# Caravan.test$.pred_Yes.BAD <- 0

Caravan.test_predictions_NULL |>
  conf_mat(truth = Purchase, estimate = .pred_class)

# But as we can see, we have no specificity!
Caravan.test_predictions_NULL |>
  mset_class(truth = Purchase, estimate = .pred_class)


# Training with Imbalance Data --------------------------------------------

knn_spec <- nearest_neighbor(
  mode = "classification",
  engine = "kknn",
  neighbors = 10
)

knn_wf <- workflow(preprocessor = rec, spec = knn_spec)

knn_fit <- fit(knn_wf, data = Caravan.train)


# Up- and Down-Sampling ---------------------------------------------------

# We can also sample our data such that we artificially achieve class balances.
# The main methods are:
# down-sampling:
#   randomly subset all the classes in the training set so that their class
#   frequencies match the least prevalent class.
# up-sampling:
#   randomly sample (with replacement) the minority class(es) to be the same
#   size as the majority class.
# hybrid methods:
#   techniques such as SMOTE and ROSE down-sample the majority class and
#   synthesize new data points in the minority class.
# See https://themis.tidymodels.org/

# We will use up- and down-sampling:

rec_up <- rec |> themis::step_upsample(Purchase)
rec_down <- rec |> themis::step_downsample(Purchase)

knn_fit.up <- knn_wf |>
  update_recipe(rec_up) |>
  fit(data = Caravan.train)
knn_fit.down <- knn_wf |>
  update_recipe(rec_down) |>
  fit(data = Caravan.train)


# Comparing Results -------------------------------------------------------

# Get raw predictions
Caravan.test_predictions <-
  bind_rows(
    "NULL" = Caravan.test_predictions_NULL,
    None = augment(knn_fit, new_data = Caravan.test),
    Up = augment(knn_fit.up, new_data = Caravan.test),
    Down = augment(knn_fit.down, new_data = Caravan.test),

    .id = "Method"
  ) |>
  mutate(
    Method = factor(Method, levels = c("NULL", "None", "Up", "Down"))
  )

# Let's look at the predictions made by the different methods for the first test
# case:
Caravan.test_predictions |>
  select(Method, Purchase, starts_with(".pred")) |>
  slice(1, .by = Method)


Caravan.test_predictions |>
  group_by(Method) |>
  mset_class(truth = Purchase, estimate = .pred_class)
# As we can see, the accuracy (and specificity) have dropped, but sensitivity is
# higher.

# We can also compare ROC curves and AUCs:
Caravan.test_predictions |>
  group_by(Method) |>
  roc_curve(truth = Purchase, .pred_Yes) |>
  autoplot()

File: py/03 classification/03 imbalanced outcome.py
import pandas as pd
from plotnine import *

from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix
from sklearn.metrics import (
    accuracy_score,
    recall_score,
    roc_curve,
    roc_auc_score,
)
from sklearn.neighbors import KNeighborsClassifier
from sklearn.dummy import DummyClassifier
from sklearn.preprocessing import StandardScaler
from sklearn import set_config

from imblearn.under_sampling import RandomUnderSampler
from imblearn.over_sampling import RandomOverSampler
from imblearn.pipeline import Pipeline  # we need a different pipe!

from ISLP import load_data

set_config(display="diagram")


# The Data & Problem --------------------------------------------------------

Caravan = load_data("Caravan")
print(Caravan.info())


# Data Splitting
outcome = "Purchase"
features = Caravan.columns.difference(["Purchase"])  # all but the outcome

X_train, X_test, y_train, y_test = train_test_split(
    Caravan[features],
    Caravan[outcome],
    train_size=0.7,
    random_state=20251202,
    stratify=Caravan[outcome],
)


print(y_train.value_counts(normalize=True))
# As we can see, the classes are very unbalanced.

# Note we used a stratified split, so that both the train and test set have
# about the ~same distribution of classes. This is particularly important with
# imbalanced data.
print(y_test.value_counts(normalize=True))


# The worst model ---------------------------------------------------------
# The data imbalance means that, technically, we can achieve high accuracy by
# simply predicting "No"....

dummy_model = DummyClassifier().fit(X_train, y_train)
# The dummy model is the worst possible model - it does not use any information
# in X, only the distribution of Y in the training set.
# - For regression, it always predicts mean(Y)
# - For classification, it predicts the frequent class and base rate
#   probabilities.
# These models are good for benchmarking.

# But like a "real" model, it can be used to generate predictions on new data:
y_pred_null = dummy_model.predict(X_test)
y_pred_proba_null = dummy_model.predict_proba(X_test)

# same as:
# y_pred_null = np.array(["No"] * len(y_test))
# y_pred_proba_null = np.tile(
#     y_train.value_counts(normalize=True), (len(y_test), 1)
# )

# Confusion matrix
cm_null = confusion_matrix(y_test, y_pred_null, labels=["Yes", "No"])
print("\nNull model confusion matrix:")
print(pd.DataFrame(cm_null, index=["Yes", "No"], columns=["Yes", "No"]))


# But as we can see, we have no sensitivity!
def specificity_score(y_true, y_pred):
    # Specificity = TN / (TN + FP)
    # For binary classification with pos_label, this is recall of negative class
    tn, fp, fn, tp = confusion_matrix(
        y_true, y_pred, labels=["No", "Yes"]
    ).ravel()
    return tn / (tn + fp) if (tn + fp) > 0 else 0


accuracy_null = accuracy_score(y_test, y_pred_null)
sensitivity_null = recall_score(y_test, y_pred_null, pos_label="Yes")
specificity_null = specificity_score(y_test, y_pred_null)

print("\n" + "=" * 60)
print("Null Model Performance:")
print("=" * 60)
print(f"Accuracy:    {accuracy_null:.3f}")
print(f"Sensitivity: {sensitivity_null:.3f}")
print(f"Specificity: {specificity_null:.3f}")
print("=" * 60)


# Training with Imbalance Data --------------------------------------------

knn_pipe = Pipeline(
    steps=[
        ("z", StandardScaler()),
        ("classifier", KNeighborsClassifier(n_neighbors=10)),
    ]
)

knn_pipe.fit(X_train, y_train)


# Up- and Down-Sampling ---------------------------------------------------

# We can also sample our data such that we artificially achieve class balances.
# The main methods are:
# down-sampling:
#   randomly subset all the classes in the training set so that their class
#   frequencies match the least prevalent class.
# up-sampling:
#   randomly sample (with replacement) the minority class(es) to be the same
#   size as the majority class.
# hybrid methods:
#   techniques such as SMOTE and ROSE down-sample the majority class and
#   synthesize new data points in the minority class.
# See https://imbalanced-learn.org/

# We will use up- and down-sampling:

knn_down_pipe = Pipeline(
    steps=[
        ("downsamp", RandomUnderSampler()),  # added here
        ("z", StandardScaler()),
        ("classifier", KNeighborsClassifier(n_neighbors=10)),
    ]
)

knn_up_pipe = Pipeline(
    steps=[
        ("upsamp", RandomOverSampler()),  # added here
        ("z", StandardScaler()),
        ("classifier", KNeighborsClassifier(n_neighbors=10)),
    ]
)

# Note these are imblearn.pipeline.Pipeline() and NOT
# sklearn.pipeline.Pipeline()!

knn_down_pipe.fit(X_train, y_train)
knn_up_pipe.fit(X_train, y_train)


# Comparing Results -------------------------------------------------------

# Comparing Results -------------------------------------------------------

# Get predictions from all models
y_pred_none = knn_pipe.predict(X_test)
y_pred_proba_none = knn_pipe.predict_proba(X_test)

y_pred_up = knn_up_pipe.predict(X_test)
y_pred_proba_up = knn_up_pipe.predict_proba(X_test)

y_pred_down = knn_down_pipe.predict(X_test)
y_pred_proba_down = knn_down_pipe.predict_proba(X_test)


# Calculate metrics for each method
metrics_by_method = []
for pred_class, method in zip(
    [y_pred_null, y_pred_none, y_pred_up, y_pred_down],
    ["NULL", "None", "Up", "Down"],
):
    metrics_by_method.append(
        {
            "Method": method,
            "Accuracy": accuracy_score(y_test, pred_class),
            "Sensitivity": recall_score(
                y_test,
                pred_class,
                pos_label="Yes",
            ),
            "Specificity": specificity_score(y_test, pred_class),
        }
    )

metrics_df = pd.DataFrame(metrics_by_method)
print("\n" + "=" * 60)
print("Performance Comparison:")
print("=" * 60)
print(metrics_df.to_string(index=False))
print("=" * 60)
# As we can see, the accuracy (and specificity) have dropped, but sensitivity is
# higher.


# We can also compare ROC curves and AUCs:
roc_data = []
for pred_proba, method in zip(
    [y_pred_proba_null, y_pred_proba_none, y_pred_proba_up, y_pred_proba_down],
    ["NULL", "None", "Up", "Down"],
):
    fpr, tpr, _ = roc_curve(y_test, pred_proba[:, 1], pos_label="Yes")
    roc_data.append(pd.DataFrame({"fpr": fpr, "tpr": tpr, "Method": method}))

roc_df = pd.concat(roc_data)

p_roc = (
    ggplot(roc_df, aes(x="fpr", y="tpr", color="Method"))
    + geom_line(size=1)
    + geom_abline(intercept=0, slope=1, linetype="dashed", color="gray")
    + labs(
        x="False Positive Rate (1 - Specificity)",
        y="True Positive Rate (Sensitivity)",
        title="ROC Curves by Sampling Method",
        color="Method",
    )
    + theme_minimal()
    + coord_equal()
    + xlim(0, 1)
    + ylim(0, 1)
)
p_roc.draw(show=True)

# Calculate AUC for each method
for pred_proba, method in zip(
    [y_pred_proba_null, y_pred_proba_none, y_pred_proba_up, y_pred_proba_down],
    ["NULL", "None", "Up", "Down"],
):
    auc = roc_auc_score(y_test, pred_proba[:, 1])
    print(f"  {method}: {auc:.3f}")
